R/utils.R

```diff
@@ -283,22 +283,78 @@ as_slide_computation <- function(f, ...) {
       data_mask$.ref_time_value <- .ref_time_value
-      rlang::eval_tidy(f, data_mask)
+      common_size <- NULL
+      results_names <- character(0L) # track ordering; env doesn't
+      for (quosure_i in seq_along(f)) {
+        # XXX could capture and improve error messages here at cost of recover()ability
+        quosure_result_raw <- rlang::eval_tidy(quosures[[quosure_i]], data_mask)
+        if (is.null(quosure_result_raw)) {
+          nm <- nms[[quosure_i]]
+          results_names <- results_names[results_names != nm]
+          rlang::env_unbind(results_env, nm)
+        } else if (
+          # vctrs considers data.frames to be vectors, but we still check
+          # separately for them because certain base operations output data frames
+          # with rownames, which we will allow (but might drop)
+          is.data.frame(quosure_result_raw) ||
+            vctrs::obj_is_vector(quosure_result_raw) && is.null(vctrs::vec_names(quosure_result_raw))
+        ) {
+          # We want something like `dplyr_col_modify()` but allowing recycling
+          # of previous computations and updating `results_env` and unpacking
+          # tibbles if not manually named.
+          if (!is.null(common_size)) {
+            # XXX could improve error messages here
+            quosure_result_recycled <- vctrs::vec_recycle(quosure_result_raw, common_size)
+          } else {
+            quosure_result_recycled <- quosure_result_raw
+            quosure_result_size <- vctrs::vec_size(quosure_result_raw)
+            if (quosure_result_size != 1L) {
+              common_size <- quosure_result_size
+              for (previous_result_nm in names(results_env)) {
+                results_env[[previous_result_nm]] <- vctrs::vec_recycle(results_env[[previous_result_nm]], common_size)
+              }
+            } # else `common_size` remains NULL
+          }
+          if (inherits(quosure_result_recycled, "data.frame") && !manually_named[[quosure_i]]) {
+            new_results_names <- names(quosure_result_recycled)
+            results_names <- c(results_names, new_results_names)
+            for (new_result_i in seq_along(quosure_result_recycled)) {
+              results_env[[new_results_names[[new_result_i]]]] <- quosure_result_recycled[[new_result_i]]
+            }
+          } else {
+            nm <- nms[[quosure_i]]
+            results_names <- c(results_names, nm)
+            results_env[[nm]] <- quosure_result_recycled
+          }
+        } else {
+          cli_abort("
+            Problem with output of {.code
+            {rlang::expr_deparse(rlang::quo_get_expr(f[[quosure_i]]))}}; it
+            produced a result that was neither NULL, a data.frame, nor a vector
+            without unnamed entries (as determined by the vctrs package).
+          ", class = "epiprocess__invalid_slide_comp_tidyeval_output")
+        }
+      }
+      validate_tibble(new_tibble(as.list(results_env, all.names = TRUE)[results_names]))
```

Comment:
issue?: in rare case that user redefines a col within a computation, we'll request same colname twice here.  Somehow gets de-duped down the line, but probably better to make some comments or fixes here.

Reply:
should be done